version: '3'

services:

  backend:
    container_name: backend
    image: django
    pull_policy: never
    build: ./backend
    env_file:
      - .env
    depends_on:
      users_db:
        condition: service_healthy
      game_db:
        condition: service_healthy
    ports:
      - "8000:8000"
    volumes:
      - backend_vol:/var/www/django/backend/
    networks:
      - pong_net

  users_db:
    container_name: users_db
    image: postgres:14.11-alpine3.18
    restart: always
    environment:
    # todo POSTGRES_DB 연결 안 돼서 User 이름과 동일하게 설정 함.
    # https://hub.docker.com/_/postgres
      POSTGRES_USER: ${USERS_DB_USER}
      POSTGRES_PASSWORD: ${USERS_DB_PW}
      POSTGRES_DB: ${USERS_DB_NAME}
    volumes:
      - users_db_vol:/var/lib/postgresql/data
    expose:
      - "5432"
    networks:
      - pong_net
    healthcheck:
      test: ["CMD-SHELL", "pg_isready", "-d", "db_prod"]
      interval: 10s
      timeout: 5s
      retries: 5

  game_db:
    container_name: redis
    image: redis:alpine
    restart: always
    command: redis-server --port 6379
    volumes:
      - game_db_vol:/data
    expose:
      - "6379"
    networks:
      - pong_net
    healthcheck:
      test: [ "CMD", "redis-cli", "--raw", "incr", "ping" ]
      interval: 10s
      timeout: 5s
      retries: 5

networks:
  pong_net:
    driver: bridge

volumes:
# 볼륨
  users_db_vol:
  game_db_vol:
# 바인드 마운트
  backend_vol:
    driver: local
    driver_opts:
      o: bind
      type: none
      device: ${BACKEND_VOL_PATH}
